src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp

```diff
@@ -378,88 +421,101 @@ FixedwingPositionControl::calculate_target_airspeed(float airspeed_demand, const
-	// sanity check: limit to range
-	return constrain(airspeed_demand, adjusted_min_airspeed, _param_fw_airspd_max.get());
+	return constrain(airspeed_setpoint, adjusted_min_airspeed, _param_fw_airspd_max.get());
 }
 
 void
-FixedwingPositionControl::tecs_status_publish()
+FixedwingPositionControl::updateSpeedMode()
 {
-	tecs_status_s t{};
+	FW_SPEED_MODE_COMMANDED new_mode = static_cast<FW_SPEED_MODE_COMMANDED>(_fw_spd_mode_set.get());
 
-	switch (_tecs.tecs_mode()) {
-	case TECS::ECL_TECS_MODE_NORMAL:
-		t.mode = tecs_status_s::TECS_MODE_NORMAL;
+	switch (new_mode) {
+	case NORMAL:
+		_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_NORMAL; //default
 		break;
 
-	case TECS::ECL_TECS_MODE_UNDERSPEED:
-		t.mode = tecs_status_s::TECS_MODE_UNDERSPEED;
-		break;
+	case ECO_CRUISE:
+		if (_conditions_for_eco_dash_met && _tecs.get_flight_phase() == tecs_status_s::TECS_FLIGHT_PHASE_LEVEL) {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_ECO;
+
+		} else {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_NORMAL;
+		}
 
-	case TECS::ECL_TECS_MODE_BAD_DESCENT:
-		t.mode = tecs_status_s::TECS_MODE_BAD_DESCENT;
 		break;
 
-	case TECS::ECL_TECS_MODE_CLIMBOUT:
-		t.mode = tecs_status_s::TECS_MODE_CLIMBOUT;
+	case ECO_FULL:
+		if (_conditions_for_eco_dash_met) {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_ECO;
+
+		} else {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_NORMAL;
+		}
+
 		break;
-	}
 
-	t.altitude_sp = _tecs.hgt_setpoint();
-	t.altitude_filtered = _tecs.vert_pos_state();
+	case DASH_CRUISE:
+		if (_conditions_for_eco_dash_met && _tecs.get_flight_phase() == tecs_status_s::TECS_FLIGHT_PHASE_LEVEL) {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_DASH;
 
-	t.true_airspeed_sp = _tecs.TAS_setpoint_adj();
-	t.true_airspeed_filtered = _tecs.tas_state();
+		} else {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_NORMAL;
+		}
 
-	t.height_rate_setpoint = _tecs.hgt_rate_setpoint();
-	t.height_rate = _tecs.vert_vel_state();
+		break;
 
-	t.equivalent_airspeed_sp = _tecs.get_EAS_setpoint();
-	t.true_airspeed_derivative_sp = _tecs.TAS_rate_setpoint();
-	t.true_airspeed_derivative = _tecs.speed_derivative();
-	t.true_airspeed_derivative_raw = _tecs.speed_derivative_raw();
-	t.true_airspeed_innovation = _tecs.getTASInnovation();
+	case DASH_FULL:
+		if (_conditions_for_eco_dash_met) {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_DASH;
 
-	t.total_energy_error = _tecs.STE_error();
-	t.total_energy_rate_error = _tecs.STE_rate_error();
+		} else {
+			_speed_mode_current = FW_SPEED_MODE::FW_SPEED_MODE_NORMAL;
+		}
 
-	t.energy_distribution_error = _tecs.SEB_error();
-	t.energy_distribution_rate_error = _tecs.SEB_rate_error();
+		break;
+	}
+}
 
-	t.total_energy = _tecs.STE();
-	t.total_energy_rate = _tecs.STE_rate();
-	t.total_energy_balance = _tecs.SEB();
-	t.total_energy_balance_rate = _tecs.SEB_rate();
+void
+FixedwingPositionControl::check_eco_dash_allowed()
```

Comment:
```suggestion
FixedwingPositionControl::updateEcoDashAllowed()
```

Reply:
I've actually removed check_eco_dash_allowed() and check it directly inside getSpeedMode(). Have a look at https://github.com/PX4/PX4-Autopilot/pull/18605/commits/aeda4b6546aada6dcaee5e76eecba5d0665ac36a